README.md
# PyBozoCrack
PyBozoCrack is a depressingly effective MD5 password hash cracker with almost zero CPU/GPU load. Instead of rainbow tables, dictionaries, or brute force, BozoCrack simply *finds* the plaintext password. Specifically, it googles the MD5 hash and hopes the plaintext appears somewhere on the first page of results.

It works way better than it ever should.


## How?
Basic usage:

    $ python bozocrack.py -f my_md5_hashes.txt

Or:

    $ python bozocrack.py -s fcf1eed8596699624167416a1e7e122e

The input file has no specified format. BozoCrack automatically picks up strings that look like MD5 hashes. A single line shouldn't contain more than one hash.

Example with output:

    $ python bozocrack.py -f example.txt
    Loaded 5 unique hashes
    fcf1eed8596699624167416a1e7e122e:octopus
    bed128365216c019988915ed3add75fb:passw0rd
    d0763edaa9d9bd2a9516280e9044d885:monkey
    dfd8c10c1b9b58c8bf102225ae3be9eb:12081977
    ede6b50e7b5826fe48fc1f0fe772c48f:1q2w3e4r5t6y

    $ python bozocrack.py -s fcf1eed8596699624167416a1e7e122e
    fcf1eed8596699624167416a1e7e122e:octopus


## Why?
To show just how bad an idea it is to use plain MD5 as a password hashing mechanism. Honestly, if the passwords can be cracked with *this software*, there are no excuses.


## Who?
BozoCrack was originally written by [Juuso Salonen](http://twitter.com/juusosalonen), the guy behind [Radio Silence](http://radiosilenceapp.com) and [Private Eye](http://radiosilenceapp.com/private-eye). BozoCrack was rewritten in Python by [Henrique Pereira](http://twitter.com/ikkebr).


## License
Do whatever you wish. Public domain, yadda yadda.


HISTORY.rst
.. :changelog:

History
-------

1.0.2 (2014-05-02)
++++++++++++++++++

* Increased test coverage

1.0.1 (2014-05-01)
++++++++++++++++++

* Trying to fix problems with Travis

1.0 (2014-05-01)
++++++++++++++++++

* First release on PyPI.

tests/test_pybozocrack.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_pybozocrack
----------------------------------

Tests for `pybozocrack` module.
"""

import unittest

from pybozocrack import pybozocrack


class TestPybozocrack(unittest.TestCase):
	
    def setUp(self):
        self.hash = "d0763edaa9d9bd2a9516280e9044d885"
        self.plaintext = "monkey"

        file = open('test', 'w')
        file.write('fcf1eed8596699624167416a1e7e122e\nbed128365216c019988915ed3add75fb')
        file.close()
		
        file = open('cache', 'w')
        file.write('1:2\n')
        file.close()

        self.cracker = pybozocrack.BozoCrack('test')


    def test_loaded_hashes(self):
        self.assertEqual(len(self.cracker.hashes), 2)

    def test_load_empty_cache(self):
        self.assertEqual(self.cracker.load_cache('empty'), {})

    def test_append_to_cache(self):
        self.cracker.append_to_cache('1', '2', 'cache')
    #    self.assertEqual(self.cracker.load_cache('cache'), {'1': '2'})
		
    def test_crack(self):
        self.cracker.hashes = [self.hash,]
        self.cracker.crack()
        self.assertEqual( self.cracker.cache[self.cracker.hashes[0]], self.plaintext )
        
    def test_dictionary_attack_known_hash(self):
        self.assertEqual(pybozocrack.dictionary_attack(self.hash, ['zebra', '123', self.plaintext]), self.plaintext)
		
    def test_dictionary_attack_invalid_hash(self):
        self.assertIsNone(pybozocrack.dictionary_attack(self.hash, ['zebra', '123']))
		
    def test_format_it(self):
        self.assertEqual(pybozocrack.format_it(self.hash, self.plaintext), "{}:{}".format(self.hash, self.plaintext))

    def test_crack_single_hash(self):
        self.assertEqual(pybozocrack.crack_single_hash(self.hash), self.plaintext)

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


bozocrack.rb
#!/usr/bin/env ruby

require 'digest/md5'
require 'net/http'

class BozoCrack

  def initialize(filename)
    @hashes = Array.new
    @cache = Hash.new

    File.new(filename).each_line do |line|
      if m = line.chomp.match(/\b([a-fA-F0-9]{32})\b/)
        @hashes << m[1]
      end
    end
    @hashes.uniq!
    puts "Loaded #{@hashes.count} unique hashes"

    load_cache
  end

  def crack
    @hashes.each do |hash|
      if plaintext = @cache[hash]
        puts "#{hash}:#{plaintext}"
        next
      end
      if plaintext = crack_single_hash(hash)
        puts "#{hash}:#{plaintext}"
        append_to_cache(hash, plaintext)
      end
      sleep 1
    end
  end

  private

  def crack_single_hash(hash)
    if plaintext = crack_single_hash_with_website(hash, "http://www.google.com/search?q=#{hash}")
      return plaintext
    end
    if plaintext = crack_single_hash_with_website(hash, "http://www.google.com/search?q=md5+#{hash}")
      return plaintext
    end
    nil
  end

  def crack_single_hash_with_website(hash, url)
    response = Net::HTTP.get URI(url)
    wordlist = response.split(/\s+/)
    if plaintext = dictionary_attack(hash, wordlist)
      return plaintext
    end
    nil
  end

  def dictionary_attack(hash, wordlist)
    wordlist.each do |word|
      if Digest::MD5.hexdigest(word) == hash.downcase
        return word
      end
      sub_wordlist = word.split(/[^a-zA-Z0-9]+/)
      if (sub_wordlist.size > 1)
        if plaintext = dictionary_attack(hash, sub_wordlist)
          return plaintext
        end
      end
    end
    nil
  end

  def load_cache(filename = "cache")
    if File.file? filename
      File.new(filename).each_line do |line|
        if m = line.chomp.match(/^([a-fA-F0-9]{32}):(.*)$/)
          @cache[m[1]] = m[2]
        end
      end
    end
  end

  def append_to_cache(hash, plaintext, filename = "cache")
    File.open(filename, "a") do |file|
      file.write "#{hash}:#{plaintext}\n"
    end
  end

end

if ARGV.size == 1
  BozoCrack.new(ARGV[0]).crack
else
  puts "Usage example: ruby bozocrack.rb file_with_md5_hashes.txt"
end


bozocrack.php
<html>
<head>
<title>BozoCrackPHP Simply Scary Hash Cracker</title>
</head>
<body>
<h1>BozoCrackPHP Simply Scary Hash Cracker</h1>
<?

//Check if hashes have been submitted
if(isset($_POST['hashes'])){
  //If they have been, start processing them

  //Split input into array of hashes
  $hashes = explode("\n", $_POST['hashes']);

  $hashes = valid_hashes($hashes);

  $number_of_unique_hashes = count($hashes);
  echo "Loaded $number_of_unique_hashes unique hashes<br \>\n";

  //Begin cracking hashes
  foreach($hashes as $hash){
    //If cracking is successful, output the answer
    if($plaintext = crack_single_hash($hash)){
      echo "$hash:$plaintext<br />\n";
    }
  }
}

//Otherwise, only display the submission form

?>

<form action="bozocrack.php" method="post">
<textarea name="hashes" cols=40 rows=10></textarea><br />
<input type="submit" value="Submit Hashes" />
</form>

</body>
</html>
<?
//Functions

//valid_hashes(array $hashes) takes an array of hashes and returns only valid hashes
function valid_hashes(array $hashes){
  //Make sure the hashes are unique
  $hashes = array_unique($hashes);

  $valid_hashes = array();

  //start looping through hashes
  foreach($hashes as $hash){
    //Filter special chars
    $hash = preg_replace("/[^0-9a-zA-Z]/i", '', $hash);

    //Make sure hash is 32 characters
    if(strlen($hash) == 32){
      $valid_hashes[] = $hash;
    }
    //otherwise, warn the user
    else{
      echo "Warning: $hash is not 32 characters - MD5 hash must be 32 characters long.<br />\n";
    }
  }
  return $valid_hashes;
}

//crack_single_hash($hash) returns cracked hash or false
function crack_single_hash($hash){
  //search for the hash on Google
  $response = file_get_contents("http://www.google.com/search?q=$hash");

  //split Google's response into separate words
  $wordlist = preg_split("/\s+/", $response);

  //If hash can be cracked, return the answer
  if($plaintext = dictionary_attack($hash, $wordlist)){
    return $plaintext;
  }
  return false;
}

//dictionary_attack($hash, $wordlist) if it finds the cracked hash in the wordlist, it returns it
function dictionary_attack($hash, $wordlist){
  foreach($wordlist as $word){
    if(md5($word) == strtolower($hash)){
      return $word;
    }
  }
}
?>


pybozocrack/pybozocrack.py
#!/usr/bin/env python
import hashlib
import re
from urllib import FancyURLopener
import sys
from optparse import OptionParser

HASH_REGEX = re.compile("([a-fA-F0-9]{32})")


class MyOpener(FancyURLopener):
    version = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; it; rv:1.8.1.11) Gecko/20071127 Firefox/2.0.0.11'


def dictionary_attack(h, wordlist):
    for word in wordlist:
        if hashlib.md5(word).hexdigest() == h:
            return word

    return None


def format_it(hash, plaintext):
    return "{hash}:{plaintext}".format(hash=hash, plaintext=plaintext)


def crack_single_hash(h):
    myopener = MyOpener()
    response = myopener.open(
        "http://www.google.com/search?q={hash}".format(hash=h))

    wordlist = response.read().replace('.', ' ').replace(
        ':', ' ').replace('?', '').split(' ')
    plaintext = dictionary_attack(h, set(wordlist))

    return plaintext


class BozoCrack(object):

    def __init__(self, filename, *args, **kwargs):
        self.hashes = []

        with open(filename, 'r') as f:
            hashes = [h.lower() for line in f if HASH_REGEX.match(line)
                      for h in HASH_REGEX.findall(line.replace('\n', ''))]

        self.hashes = sorted(set(hashes))

        print "Loaded {count} unique hashes".format(count=len(self.hashes))

        self.cache = self.load_cache()

    def crack(self):
        for h in self.hashes:
            if h in self.cache:
                print format_it(h, self.cache[h])
                continue

            plaintext = crack_single_hash(h)

            if plaintext:
                print format_it(h, plaintext)
                self.cache[h] = plaintext
                self.append_to_cache(h, plaintext)

    def load_cache(self, filename='cache'):
        cache = {}
        with open(filename, 'r') as c:
            for line in c:
                hash, plaintext = line.replace('\n', '').split(':', 1)
                cache[hash] = plaintext
        return cache

    def append_to_cache(self, h, plaintext, filename='cache'):
        with open(filename, 'a+') as c:
            c.write(format_it(hash=h, plaintext=plaintext)+"\n")

if __name__ == '__main__':

    parser = OptionParser()
    parser.add_option('-s', '--single', metavar='MD5HASH',
                      help='cracks a single hash', dest='single', default=False)
    parser.add_option('-f', '--file', metavar='HASHFILE',
                      help='cracks multiple hashes on a file', dest='target',)

    options, args = parser.parse_args()

    if not options.single and not options.target:
        parser.error("please select -s or -f")
    elif options.single:
        plaintext = crack_single_hash(options.single)

        if plaintext:
            print format_it(hash=options.single, plaintext=plaintext)
    else:
        BozoCrack(options.target).crack()


bozocrack.py
#!/usr/bin/env python
import hashlib
import re
from urllib import FancyURLopener
import sys
from optparse import OptionParser

HASH_REGEX = re.compile("([a-fA-F0-9]{32})")


class MyOpener(FancyURLopener):
    version = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; it; rv:1.8.1.11) Gecko/20071127 Firefox/2.0.0.11'


def dictionary_attack(h, wordlist):
    for word in wordlist:
        if hashlib.md5(word).hexdigest() == h:
            return word

    return None


def format_it(hash, plaintext):
    return "{hash}:{plaintext}".format(hash=hash, plaintext=plaintext)


def crack_single_hash(h):
    myopener = MyOpener()
    response = myopener.open(
        "http://www.google.com/search?q={hash}".format(hash=h))

    wordlist = response.read().replace('.', ' ').replace(
        ':', ' ').replace('?', '').split(' ')
    plaintext = dictionary_attack(h, set(wordlist))

    return plaintext


class BozoCrack(object):

    def __init__(self, filename, *args, **kwargs):
        self.hashes = []

        with open(filename, 'r') as f:
            hashes = [h.lower() for line in f if HASH_REGEX.match(line)
                      for h in HASH_REGEX.findall(line.replace('\n', ''))]

        self.hashes = sorted(set(hashes))

        print "Loaded {count} unique hashes".format(count=len(self.hashes))

        self.cache = self.load_cache()

    def crack(self):
        for h in self.hashes:
            if h in self.cache:
                print format_it(h, self.cache[h])
                continue

            plaintext = crack_single_hash(h)

            if plaintext:
                print format_it(h, plaintext)
                self.cache[h] = plaintext
                self.append_to_cache(h, plaintext)

    def load_cache(self, filename='cache'):
        cache = {}
        with open(filename, 'a+') as c:
            for line in c:
                hash, plaintext = line.replace('\n', '').split(':', 1)
                cache[hash] = plaintext
        return cache

    def append_to_cache(self, h, plaintext, filename='cache'):
        with open(filename, 'a+') as c:
            c.write(format_it(hash=h, plaintext=plaintext)+"\n")

if __name__ == '__main__':

    parser = OptionParser()
    parser.add_option('-s', '--single', metavar='MD5HASH',
                      help='cracks a single hash', dest='single', default=False)
    parser.add_option('-f', '--file', metavar='HASHFILE',
                      help='cracks multiple hashes on a file', dest='target',)

    options, args = parser.parse_args()

    if not options.single and not options.target:
        parser.error("please select -s or -f")
    elif options.single:
        plaintext = crack_single_hash(options.single)

        if plaintext:
            print format_it(hash=options.single, plaintext=plaintext)
    else:
        BozoCrack(options.target).crack()
